### DISCUSSION TITLE
Add a "Pre-Parse" method to server.py workflow

### DISCUSSION THREAD

**Discussion by mfbutner** [PrairieLearn/PrairieLearn#7546] [Ideas]
I think it would be nice to add a "pre-parse" method to the server.py workflow where all you do is modify raw student input **before** any element sees it. I think this would be useful for three reasons.

1. All of the times when I have had to use the current `parse` method it is because I needed to just massage the student input a little to allow for more flexibility in input. For example, I have a question where I ask students to convert hex to binary and therefore a `<pl-integer-input>` made the most sense. However, when the student inputs the answer I'm guessing that they might put whitespace between their group of 4 bits when entering their answer. `<pl-integer-input>` doesn't support interior whitespace by default so I had to remove it in the parse method. 
2. When implementing the parse method, you have to set `data["submitted_answers"]["NAME"]` to be internal representation that the element wants. These internal representations are intentionally and reasonably opaque. Some of these internal representations can be easy to guess (`pl-string-input`', `pl-integer-input`, etc.) but others I would have no idea on what to do without looking at their source code. Given that I'm just modifying the raw string input, this puts an extra hurdle in the way of writing a correctly functioning parse method. Also just dealing with input as string makes it easy to understand what we are working with and don't need to know the inner details of the element's representations. 
3. The user has to revalidate modified input. The current parse method runs after the element has attempted to process and validate the raw input. If the original raw input was invalid according to the element, which is guaranteed when the input had to be massaged, I'm stuck having to revalidate the input. This could be very tough to do correctly for the more complex elements and I'm also having to repeat work they've already done since I can't ask an element to revalidate input. I'm also likely to forget to check something that the element checks for, which will allow more bugs to slip through my questions.

**Comment by nwalters512** [comment]
One potentially hairy thing to consider when modifying `raw_sumitted_answers` is that we use that to fill in the question inputs when the page reloads. If you transform something like `0011 0101` into `00110101`, the student will see the latter value instead of what they actually entered. Maybe this is fine; just raising for discussion.

In this _specific_ situation I think the objectively best thing to do is to add support for something like `ignore-whitespace="true"` to `pl-integer-input`, which would be relatively straightforward to do.

**Reply by eliotwrobson** [reply]
Will tack that onto the current integer input PR

**Reply by mfbutner** [reply]
I'd like them to continue seeing what they entered as well but have the transformed input be sent to element. I'm guessing though that adding that would break a lot of things and require much larger changes. 

For my particular problem I also had to strip out any leading `0b`'s from it as it didn't sound like that syntax was supported when providing an explicit base. Maybe since I'm treating their answer so much like a string I should just use `pl-string-input` and just override the grade method instead to support some of the variability in answers? 

I'm not sure if you have any data on this, but when people are implementing the parse method are they generally just doing some touch ups on the raw input or are they doing a lot more complicated and varying things?

**Reply by nwalters512** [reply]
Yeah, I don't want to introduce `raw_raw_submitted_answers` and so on. One _could_ teach elements to read from `data["submitted_answers"]` if its key is already present, but I'm not at all sure that's a good idea.

`pl-string-input` might indeed be a better fit here. We don't have any way to analyze the "type" of parse functions in aggregate. I spot-checked a few, and checking for `pl-string-input` formatting (e.g. "is it a valid URL") occurs multiple times.

**Reply by jonatanschroeder** [reply]
I think a custom element derived from pl-integer-input may be an interesting option here as well.

**Reply by mfbutner** [reply]
@jonatanschroeder I did consider doing this but then I would have to copy the entire code base of `pl-integer-input` into my custom element to then add my 1 extra line of code. It's not that the copy and pasting is hard, it just feels wrong as it would be a large amount of code duplication. I'm now also disconnected from the original element, so any updates or improvements it receives I don't receive until I manually copy the new code from the original element again. Is there something I'm missing that would make it easy to derive a new element from an existing one in a "class inheritance" sense?

**Reply by eliotwrobson** [reply]
@mfbutner I just remembered this, but can you try submitting integers with underscores instead of spaces? I think that syntax might be supported.

**Reply by mfbutner** [reply]
@eliotwrobson I just tried it and underscores are supported out of the box. I didn't expect that and I don't think students would either, but of course I can add it to the instructions for that problem. Mentioning that underscores are supported in the ? help pop up would be a nice thing to add though I suspect most people will still likely miss it as it is something extra you have to click on.

**Reply by eliotwrobson** [reply]
Yes, underscore support wasn't explicitly done actually, it's just a carryover from the fact that they're natively supported in Python (I doubt the original element writer even knew). I'll add a note about it in the help popover and the docs somewhere. I actually think it's better than adding spaces, since it requires less code.
